net/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+}
+
+pub struct Handler {
+    pub handler: Option<ReceiveClosure>,
+}
+
+impl Handler {
+    fn handle(
+        &mut self,
+        from: &SerializedAddress,
+        message: &SerializedMessage,
+    ) -> Result<SerializedMessage, Error> {
+        match self.handler {
+            None => bail!("fish"),
+            Some(ref mut handler) => (handler)(from, message),
+        }
+    }
+}
+
+pub trait Transport {
+    /** initialize the transport
+     * this might be used for example in a TCP based transport to specify a listening port for
+     * for the transport.
+     */
+    fn initialize(&mut self, config: Option<String>) -> Result<(), Error>;
+
+    /** return a default config with best configuration guesses for the transport
+     */
+    fn get_default_config(&self) -> String;
+
+    /** register a peer as a node in the transport
+     */
+    fn new_node(&mut self, addr: SerializedAddress, handler: Option<Handler>) -> Result<(), Error>;
+
+    /** send a message to a node over the transport
+     * assumes that the sending address was registered locally with new_node
+     */
+    fn send(
+        &mut self,
+        from: &SerializedAddress,
+        to: &SerializedAddress,
+        msg: SerializedMessage,
+        callback: SendResponseClosure,
+    ) -> Result<(), Error>;
+
+    /** deliver a message to a given recipient
+     * assumes that recipient address was registered locally with new_node
+     * this function may be called by the routing layer to bridge across transports
+     */
+    fn deliver(
+        &mut self,
+        from: &SerializedAddress,
+        to: &SerializedAddress,
+        message: SerializedMessage,
+    ) -> Result<SerializedMessage, Error>;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{collections::HashMap, str};
+    //    use error::NetworkError;
+
+    pub struct SimpleNode {
+        hc_addr: SerializedAddress,
+        transport_addr: u32,
+    }
+
+    impl Node for SimpleNode {
+        fn get_address(&self) -> &SerializedAddress {
+            &self.hc_addr
+        }
+
+        fn get_transport_address(&self) -> TransportAddress {
+            format!("{}", self.transport_addr).into()
+        }
+    }
+
+    pub struct SimpleTransport {
+        config: String,
+        nodes: Vec<Arc<SimpleNode>>,
+        handlers: HashMap<SerializedAddress, Handler>,
+    }
+
+    impl SimpleTransport {
+        pub fn new() -> SimpleTransport {
+            SimpleTransport {
+                config: "".into(),
+                nodes: Vec::new(),
+                handlers: HashMap::new(),
+            }
+        }
+        pub fn exists(&self, addr: &SerializedAddress) -> bool {
+            if let Some(_node) = self.nodes.iter().find(|node| *node.get_address() == *addr) {
+                true
+            } else {
+                false
+            }
+        }
+    }
+
+    fn to_str(vec: &Vec<u8>) -> String {
+        str::from_utf8(vec).unwrap().to_string()
```

Comment:
Please use `String::from_utf8` here

Reply:
can't because then I can't borrow because String::from_utf8 takes a Vec not a &Vec.